# Compositional differential abundance testing with ALDEx2 with Moving Pictures dataset

ALDEx2 is a differential abundance package that was initially developed for meta-RNA-Seq, but has performed very well with traditional RNA-Seq, 16S rRNA gene sequencing, and selective growth-type (SELEX) experiments. In principle, ALDEx2 is generalizable and can be applied to nearly any type of data that is generated by high-throughput sequencing that generates tables of per-feature counts for each sample [Fernandes 2014](https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-15). Here we will apply it to the "Moving Pictures" dataset by [Caporaso et al. 2011](https://www.ncbi.nlm.nih.gov/pubmed/21624126) to investigate differential abundance of the gut in two individuals.

We will start with the artifact generated by the DADA2 pipeline, `table.qza`. The metadata can be downloaded in the first step in the [Moving Pictures tutorial](https://docs.qiime2.org/2019.7/tutorials/moving-pictures/). The first step for ALDEx2 is to filter the samples to only compare the gut site.

```
qiime feature-table filter-samples \
  --i-table table.qza \
  --m-metadata-file sample-metadata.tsv \
  --p-where "[body-site]='gut'" \
  --o-filtered-table gut-table.qza
```

The next step will be to run ALDEx2. The full pipeline is implemented in the `aldex2` function The input is the `FeatureTable[Frequency]`, as well as a metadata file. The metadata file is necessary for defining the different groups you will be testing. For this tutorial, the groups are identified by the subject column from the metadata file. ALDEx2 automatically adds a pseudocount to remove zeros from the data, and filters any samples with 0 reads.

```
qiime aldex2 aldex2 --i-table gut-table.qza --m-metadata-file gut-metadata.tsv --p-condition subject --output-dir gut-test
```

The output artifact is `differentials.qza`, which contains a summary of the ALDEx2 output (difference, dispersion, effect, q-score, etc). From this artifact, we can visualize and extract the differentially abundant features. It is important to visualize the size of the difference between conditions (difference) as well as the size of the difference within conditions (dispersion) to capture the full context of the within-group variation. One feature may appear as differentially expressed if it has a very small dispersion and slightly larger difference, while another may have a large difference, but an even larger dispersion. These are both cases where caution should be used when calling differentially expressed features.

The visualizer `aldex2 effect-plot` takes as input the `differentials.qzv` artifact, and creates several plots (interactivity will be added soon, I promise!!!). More information about these plots can be found [in the ALDEx2 vignette](http://bioconductor.org/packages/release/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.pdf).

```
qiime aldex2 effect-plot --i-table gut-test/differentials.qza --o-visualization gut-test/gut_test
```

The plots are then viewed using the `qiime tools view` command:

```
qiime tools view gut_test.qzv
```

Yay! We have differentially expressed features (coloured in red). We can extract the features and detailed information from the ALDEx2 summary output using `extract-differences`.

```
qiime aldex2 extract-differences --i-table gut-test/differentials.qza --o-differentials gut-test/sig_gut --p-sig-threshold 0.1 --p-effect-threshold 0 --p-difference-threshold 0
```

The tab-separated output can be viewed in a text editing program after extraction:

```
qiime tools export --input-path gut-test/sig_gut.qza --output-path differentially-expressed-features
```

# Tutorial with raw count tables

This tutorial is meant for users starting from scratch.

Install qiime2 and activate your conda environment ([see qiime2 Getting Start](https://docs.qiime2.org/2019.7/getting-started/)).

Install this github repo:
```
# clone to working directory for qiime
git clone https://github.com/dgiguer/q2-aldex2 ./q2-aldex2
cd ./q2-aldex2
pip install -e .
cd ../
```

For this, we will use the example dataset provided in the ALDEx2 R package. For users of R, this a good example of how to prepare your counts table and metadata for importing into `qiime2`.

Open up R, load and save the example dataset.

```
library(ALDEx2)
data(selex)

# rearrange
d <- cbind(rownames(selex), selex[,1:14])

# this is needed for qiime
colnames(d)[1] <- "featureid"

rownames(d) <- NULL

# save the table
write.table(d, "selex.txt", sep = "\t", col.names=TRUE, row.names=FALSE, quote = FALSE)

# make the metadata table
# sampleid is necessary for qiime
metadata <- data.frame(sampleid = colnames(d[,2:15]), conds = c(rep("NS", 7), rep("S", 7)))

# write it to file
write.table(metadata, "selex_metadata.txt", sep = "\t", col.names=TRUE, row.names=FALSE, quote = FALSE)

#exit R

# ensure you have biom installed

# convert the selex dataset to biom2.0
biom convert -i selex.txt -o selex.hdf5 --table-type="OTU table" --to-hdf5
```

Now the count table and metadata have been prepared, we can import them into `qiime2` for visualization.

```
# import this into qiime2
# this makes an artifact
qiime tools import --input-path selex.hdf5 --type FeatureTable[Frequency] --input-format BIOMV210Format --output-path selex-table.qza

# conds is title of column in metadata to be used as experimental conditions
# this is the column used to decide which group each sample is in.
qiime aldex2 aldex2 --i-table selex-table.qza --m-metadata-file selex_metadata.txt --p-condition conds --output-dir aldex_output

# p is effect plot or MA plot. currently only effect plot is available.
qiime aldex2 effect-plot --i-table aldex_output/differentials.qza --p-type effect --o-visualization selex_effect

# visualize using the qiime tools view
# this will pop up the visualizer in a chrome window
qiime tools view selex_effect.qzv

```

This should generate several effect plots. To extract the differentially abundant points, use `extract-differences` and look at the `differentials` file by extracting it

```
qiime aldex2 extract-differences --i-table aldex_output/differentials.qza --o-differentials aldex_output/selex_significant_features --p-sig-threshold 0.1 --p-effect-threshold 0 --p-difference-threshold 0

qiime tools extract --input-pathaldex_output/selex_significant_features.qza --output-path differentially-expressed-features
```

Voila!
